torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1744,3 +1745,31 @@ def test_use_rref_after_shutdown(self):
         ):
             import torch.distributed.rpc.internal as internal
             internal.serialize(rref)
+
+@unittest.skipIf(
+    not torch._six.PY3,
+    "Pytorch distributed autograd package " "does not support python2",
+)
+class FaultyAgentRpcTest(RpcAgentTestFixture):
```

Comment:
Does it make sense if we let this class inherit all tests from `RpcTest`? So that we will automatically get a very good coverage on the retry algorithm? Or is it already doing so and I missed it?

Reply:
We currently only have those 2 tests for the faulty RPC Agent (`test_check_failing_messages` in rpc_test.py and `test_context_cleanup_tensor_with_grad` in dist_autograd_test.py). There's an earlier comment from Pritam suggesting just the failed messages test, basically as a representative scenario using most of our retryable messages, and ensuring they still work well with retries in the RRef context. The retry mechanism itself has a couple thorough C++ tests internally in ProcessGroupAgentTest.cpp

I tried inheriting all the tests RpcTest and DistAutogradTest, since it would definitely be great for test coverage - there are 4 tests that fail with RpcTest. I haven't dug too deep into all of them yet, but we do get RRef leaks if we don't wait for the delUser messages to reach, which do take a while since we're failing the first 3 sends (so close to 5 seconds of backoff plus the timeout duration for each message). So, we would need to make changes to make them pass.

Reply:
I see. Would I be correct if I assume this PR does not introduce any additional messages at all if there is no failure? The retry thread will just stay idle and not consuming any resources?

Reply:
We are actually forcing the first 3 sends to fail for all messages that use `sendWithRetries` (the 4 messages defined in the trap), so the retry thread is active when we use the faulty agent. By additional messages, do you mean message types outside of those 4?

Reply:
Sorry for being vague. I was mainly concerned about the behavior seen by normal users, where there is no or very little failures. Say I initialized RPC with PG agent, and I run a `rpc_sync`. After this change, is there any additional overhead compared to before?

Reply:
Ah I understand - if there are no failures, the retry thread will just stay idle. There should also be no added memory usage since messages are only added to the retryMap if they fail. With little to no failures, there shouldn't be additional overhead compared to before.